ecommerce/api.py

```diff
@@ -103,3 +111,184 @@ def make_reference_id(order):
+        ):
+            valid_coupons.append(cv)
+    return sorted(valid_coupons, key=lambda x: x.invoice_version.amount, reverse=True)
+
+
+def get_eligible_coupons(product, code=None):
+    """
+    Return the eligible coupon ids for a product.  This may include both valid and invalid coupons.
+    Invalid coupons should be filtered out with the get_valid_coupon_versions function.
+
+    Args:
+        product (Product): product to filter CouponEligibility by
+        code (str): A coupon code to filter by
+
+    Returns:
+        TimestampedModelQuerySet: list of coupon ids that can be used with the products.
+    """
+    query = CouponEligibility.objects.select_related("coupon").filter(
+        product=product, coupon__enabled=True
+    )
+    if code:
+        query = query.filter(coupon__coupon_code=code)
+
+    return query.values_list("coupon", flat=True)
+
+
+def best_coupon_for_basket(basket, auto_only=False, code=None):
+    """
+    Get the best eligible coupon for the basket.
+    Assumes that the basket only contains one item/product.
+
+    Args:
+        basket (Basket): the basket Object
+        auto_only (bool): Only retrieve `automatic` Coupons
+        code (str): A coupon code to filter by
+
+    Returns:
+        CouponVersion: the CouponVersion with the highest discount, or None
+    """
+    basket_item = basket.basketitems.first()
+    if basket_item:
+        return best_coupon_for_product(
+            basket_item.product, basket.user, auto_only=auto_only, code=code
+        )
+    return None
+
+
+def best_coupon_for_product(product, user, auto_only=False, code=None):
+    """
+    Get the best eligible coupon for a product and user.
+
+    Args:
+        product (Product): the Product Object
+        auto_only (bool): Only retrieve `automatic` Coupons
+        code (str): A coupon code to filter by
+
+    Returns:
+        CouponVersion: the CouponVersion with the highest product discount, or None
+    """
+    product_coupons = get_eligible_coupons(product, code=code)
+    if product_coupons:
+        validated_versions = get_valid_coupon_versions(product_coupons, user, auto_only)
+        if validated_versions:
+            return validated_versions[0]
+    return None
+
+
+def get_product_price(product):
+    """
+    Retrieve the price for the latest version of a product
+
+    Args:
+        product (Product): A product object
+
+    Returns:
+        Decimal: the price of a product
+    """
+    return product.productversions.order_by("-created_on").first().price
+
+
+def get_discount_price(coupon_version, product):
+    """
+    Determine the new discounted price for a product after the coupon discount is applied
+
+    Args:
+        coupon_version (CouponVersion): the CouponVersion object
+        product: (Product): the Product object
+
+    Returns:
+        Decimal: the discounted price for the Product
+    """
+    return Decimal(
```

Comment:
I don't think we need the surrounding `Decimal` here. The `int - Decimal` will result in a `Decimal`, and the other terms are all `Decimal`. If one of the terms was a `float` an exception would be raised.

Reply:
✔️